src/main/java/com/griddynamics/serialDeserialTask/SerialDeserial.java

```diff
@@ -0,0 +1,242 @@
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class SerialDeserial {
+    public static void main(String[] args) {
+        URL resource = SerialDeserial.class.getClassLoader().getResource("vehicle_data_serial_deserial.txt");
+        List<Vehicle> vehicles = null;
+        if (resource == null) {
+            System.out.println("Can't find input file.");
+            return;
+        }
+        try {
+            vehicles = readFromMainFile(resource.getPath());
+        } catch (FileFormatException | IOException e) {
+            e.printStackTrace();
+        }
+        if (vehicles != null && !vehicles.isEmpty()) {
+            writeIntoOneFile(vehicles, resource);
+            try {
+                readFromCreatedFile(resource);
+            } catch (IOException e) {
+                e.printStackTrace();
+            }
+        }
+    }
+
+    public static class VehicleDeserializer implements JsonDeserializer<Vehicle> {
+        private String vehicleTypeElementName;
+        private Gson gson;
+        private Map<String, Class<? extends Vehicle>> vehicleTypeRegistry;
+
+        public VehicleDeserializer(String vehicleTypeElementName) {
+            this.vehicleTypeElementName = vehicleTypeElementName;
+            this.gson = new Gson();
+            this.vehicleTypeRegistry = new HashMap<>();
+        }
+
+        public void registerGarageType(String vehicleTypeName, Class<? extends Vehicle> vehicleType) {
+            vehicleTypeRegistry.put(vehicleTypeName, vehicleType);
+        }
+
+        public Vehicle deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) {
+            JsonObject vehicleObject = json.getAsJsonObject();
+            JsonElement vehicleTypeElement = vehicleObject.get(vehicleTypeElementName);
+
+            Class<? extends Vehicle> vehicleType = vehicleTypeRegistry.get(vehicleTypeElement.getAsString());
+            return gson.fromJson(vehicleObject, vehicleType);
+        }
+    }
+
+    public static void readFromCreatedFile(URL resource) throws IOException {
+        String line = Files.readAllLines(Paths.get(resource.getPath()).getParent().resolve("new_format_vehicles.json")).get(0);
+        VehicleDeserializer deserializer = new VehicleDeserializer("typeOfVehicle");
+        deserializer.registerGarageType("CAR", Car.class);
+        deserializer.registerGarageType("TRUCK", Truck.class);
+        Gson gson = new GsonBuilder().registerTypeAdapter(Vehicle.class, deserializer).create();
+        List<Vehicle> outList = gson.fromJson(line, new TypeToken<List<Vehicle>>() {
+        }.getType());
+        for (Vehicle cur : outList) {
+            System.out.println(cur);
+        }
+    }
+
+    public static void writeIntoOneFile(List<Vehicle> vehicles, URL resource) {
+        try (Writer vehiclesWriter = new FileWriter(Paths.get(resource.getPath())
+                .getParent().resolve("new_format_vehicles.json").toFile())) {
```

Comment:
@kolegova-mary let's use new File("new_format_vehicles.json") instead of this construction "Paths.get(resource.getPath()).getParent().resolve("new_format_vehicles.json").toFile()"